gemm/src/kernels.cu

```diff
@@ -6,22 +8,56 @@ __global__ void __sgemm(const int M,
                         const int K,
                         const float* const A,
                         const float* const B,
-                        float* const C) {
+                        float* const C,
+                        const int tileDim) {
+    const int tid = threadIdx.y * blockDim.x + threadIdx.x;
+    const int block_size = blockDim.x * blockDim.y;
+
     const int i = blockIdx.x * blockDim.x + threadIdx.x;
     const int j = blockIdx.y * blockDim.y + threadIdx.y;
-    if (i >= M || j >= N)
-        return;
 
-    const float* a = A + i;          // a = A[i, 0]
-    const float* b = B + j * K;      // b = B[0, j]
     float* const c = C + i + j * M;  // c = C[i, j]
-    float sum = 0.0;
-    for (int k = 0; k < K; k++) {
-        sum += *a * *b;  // C[i, j] += A[i, k] * B[k, j]
-        a += M;          // A[i, k] -> A[i, k + 1]
-        b++;             // B[k, j] -> B[k + 1, j]
+
+    extern __shared__ float smem[];
+    float* sA = smem;                         // (blockDim.x, tileDim)
+    float* sB = smem + blockDim.x * tileDim;  // (tileDim, blockDim.y)
+
+    int gIdx;        // global memory index
+    int sIdx;        // shard memory index
+    int tIdx;        // tile index
+    float lC = 0.0;  // local memory partial summation
+
+    for (tIdx = 0; tIdx < K / tileDim + (K % tileDim != 0); tIdx++) {
+        for (sIdx = tid; sIdx < blockDim.x * tileDim; sIdx += block_size) {
+            gIdx = (tIdx * tileDim + sIdx / blockDim.x) * M +
+                   blockIdx.x * blockDim.x + sIdx % blockDim.x;
+            if (gIdx < M * K)
+                sA[sIdx] = A[gIdx];
+        }
+
+        for (sIdx = tid; sIdx < tileDim * blockDim.y; sIdx += block_size) {
+            gIdx = (blockIdx.y * blockDim.y + sIdx / tileDim) * K +
+                   tIdx * tileDim + sIdx % tileDim;
+            if (gIdx < K * N)
+                sB[sIdx] = B[gIdx];
+        }
```

Comment:
Here you're copying over memory for A and B. You aren't truncating by invalid `i` and `j` because you need those threads to copy over memory.